### PR TITLE
perf(utils): vectorise selective_log_softmax for all dtypes

### PR DESCRIPTION
**Pull request by piyushumate**
# What does this PR do?

Optimises utility function called by every alignment trainer in TRL (GRPO, RLOO, SFT, KTO, PPO) across all dtypes.

### 1. `selective_log_softmax` — vectorised both dtype paths

**fp32/fp64 path** had the same Python loop problem as bf16:

```python
# before: O(N) Python loop, one kernel dispatch per row
logsumexp_values = torch.stack([torch.logsumexp(lg, dim=-1) for lg in logits])

# after: single fully-vectorised call
logsumexp_values = torch.logsumexp(logits, dim=-1)
```

**bf16/fp16 path:** replaced the `for row_logits, row_labels in zip(...)` loop with vectorised token chunking (`chunk_size=512`), reducing O(N) kernel launches to ⌈N/512⌉. Keeps the stable half-precision strategy (no logsumexp on bf16/fp16).

## Benchmarks (NVIDIA H100 80GB HBM3, PyTorch 2.4.0)

**float32 — `selective_log_softmax`**

| Shape | Before | After | Speedup | Bit-exact |
|---|---|---|---|---|
| N=4096 V=32k (Llama-3/Mistral) | 153.94 ms | 1.15 ms | **133×** | ✅ |
| N=4096 V=128k (Llama-3.1) | 250.07 ms | 4.47 ms | **56×** | ✅ |
| N=8192 V=32k (long-context) | 312.45 ms | 2.26 ms | **138×** | ✅ |

**bfloat16 — `selective_log_softmax`**

| Shape | Before | After | Speedup | Bit-exact |
|---|---|---|---|---|
| N=4096 V=32k (Llama-3/Mistral) | 60.39 ms | 0.37 ms | **165×** | ✅ |
| N=4096 V=128k (Llama-3.1) | 114.55 ms | 1.53 ms | **75×** | ✅ |
| N=8192 V=32k (long-context) | 126.84 ms | 0.73 ms | **175×** | ✅ |

## Edge cases verified

- 3D input `(B, seq, vocab)` — all trainer call sites pass this shape ✅
- Gradient flow through `selective_log_softmax` (bf16) ✅
- Gradient flow through `entropy_from_logits` with `requires_grad=True` (GRPO entropy) ✅
- Chunk boundary correctness at N=511, 512, 513, 1024, 1025 — bit-exact ✅

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

New tests added:
- `test_selective_log_softmax_2d_flat_tokens` — covers the flat `(N, vocab)` shape that GRPO/RLOO trainers pass after reshaping (was previously untested).
- `test_selective_log_softmax_chunk_size_invariant` — verifies results are identical across chunk sizes 64/512/2048.
- All 52 tests pass; `ruff check` + `ruff format` clean.

## AI writing disclosure

- [x] AI-assisted: implementation and tests were written with AI assistance and reviewed by a human.

## Who can review?

@lewtun @qgallouedec @kashif — touches utility functions used by all alignment trainers.

---

> [!NOTE]
> **Medium Risk**
> Changes numerics implementation on a shared training utility; behavior is heavily tested and the public API only adds an optional `chunk_size`, but regressions would affect log-prob computation across trainers.
> 
> **Overview**
> **Vectorises `selective_log_softmax`** in `trl/trainer/utils.py`, a utility on the hot path for alignment trainers (log-prob gathering from logits).
> 
> For **float32/float64**, it replaces a per-row Python `logsumexp` loop with a single **`torch.logsumexp` over the full tensor**. For **bfloat16/float16**, it keeps the stable `log_softmax` strategy (no half-precision `logsumexp`) but swaps the per-row `zip` loop for a **flattened, chunked path** with a new optional **`chunk_size` (default 512)** to cap peak memory.
> 
> Tests add coverage for **2D `(N, vocab)` logits** (GRPO/RLOO-style flat tokens) and assert **bit-exact invariance** across chunk sizes 64/512/2048 on the half-precision path.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 45e7d673ed2a002187b7446ef4f8369ad93720b3. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by piyushumate** [comment]
Thanks @cursor — this was caught and fixed before the review. The in-place `probs.mul_(logps)` was reverted in commit `2da4157`. The current code uses fully out-of-place arithmetic:

```python
chunk_entropy = -(torch.exp(logps) * logps).sum(-1)
```

Gradient flow through `entropy_from_logits` with `requires_grad=True` is verified in the edge case tests — backward passes without error on both the entropy bonus path (`entropy_coef > 0`) and the no-grad logging path.

File: tests/test_utils.py [hunk 1]
```diff
@@ -934,6 +934,39 @@ def test_selective_log_softmax_multi_index(self, dtype, k):
...
+        num_tokens = 1024
+
+        input_ids = torch.randint(low=0, high=vocab_size, size=(num_tokens,))
+        logits = torch.randn(num_tokens, vocab_size, dtype=dtype)
+
+        # Reference: naive gather on log_softmax
+        expected_output = torch.gather(logits.log_softmax(-1), dim=-1, index=input_ids.unsqueeze(-1)).squeeze(-1)
+        actual_output = selective_log_softmax(logits, input_ids)
+
+        assert torch.equal(actual_output, expected_output)
```
[the remark this answers is not in the document]
Comment (piyushumate) [reply tests/test_utils.py:968]: Fixed in 45e7d67. `chunk_size` is now an explicit parameter (`default=512`) on `selective_log_softmax`, and the test passes it directly — `chunk_size=64/512/2048` — so each parametrized case exercises a genuinely different code path.
